FWCore/Framework/src/Principal.cc

```diff
@@ -581,7 +581,7 @@ namespace edm {
 
     ProductData const* result = findProductByLabel(kindOfType, typeID, inputTag, consumer, sra, mcc);
     if (result == nullptr) {
-      return BasicHandle(makeHandleExceptionFactory([=]() -> std::shared_ptr<cms::Exception> {
+      return BasicHandle(makeHandleExceptionFactory([=, *this]() -> std::shared_ptr<cms::Exception> {
```

Comment:
If I'm reading the [documentation](https://en.cppreference.com/w/cpp/language/lambda#Lambda_capture) correctly, `[=, this]` is invalid until C++20:
> If the capture-default is =, subsequent simple captures must begin with & or be *this(since C++17) or this(since C++20).

Reply:
Good point. While `[=, this]` seems to compile on GCC 12.3 with `-std=c++17`, clang issues a warning. Then, we  can either accept the deprecation warning in CPP20 build until we switch to C++20 in production and fix it then, or replace `=` with explicit by-copy capture of the non-member variables needed in the lambda body.

Reply:
I'm working on capturing everything explicitly. Or, we can capture everything by reference `[&]`, which doesn't require explicitly capturing `this`.

Reply:
You can't capture by `&` as we need the data to be copied

Reply:
> While [=, this] seems to compile on GCC 12.3 with -std=c++17, clang issues a warning. 

@makortel @Dr15Jones I think it would be easier to use `[=, this]` and use `#pragma diagnostics` to suppress `-Wc++20-extensions` in clang?